library/core/src/array/mod.rs

```diff
@@ -936,28 +941,51 @@ impl<T, const N: usize> Guard<'_, T, N> {
     }
 }
 
-impl<T, const N: usize> Drop for Guard<'_, T, N> {
+impl<T: ~const Destruct, const N: usize> const Drop for Guard<'_, T, N> {
     fn drop(&mut self) {
         debug_assert!(self.initialized <= N);
 
+        #[inline]
+        const fn drop_ct<T: ~const Destruct>(x: &mut [T]) {
+            let mut i = 0;
+            while i < x.len() {
+                // SAFETY: dropping the value, contains initialized objects
+                unsafe {
+                    crate::ptr::read(&mut x[i]);
+                }
+                i += 1;
+            }
+        }
```

Comment:
AIUI `drop_in_place` on slices has different behavior compared to a loop that drops individual items. I'm not sure if we care about that difference here, just pointing out that these two impls aren't exactly equivalent.

Reply:
https://github.com/rust-lang/rust/blob/7c75fe4c8547c276574cacb144919d67fd8ab302/compiler/rustc_mir_dataflow/src/elaborate_drops.rs#L651-L667